### ISSUE TITLE
IOU metric sometimes bigger than 1

### ISSUE THREAD

**Issue by Diyago** [qubvel/segmentation_models.pytorch#17]
I have slightly modified script from car segmentation ipynb file for my own binary segmentation mask:

If I use same training loop for resnet 18 I get very high IUO even higher than 1 which impossible.

Using evaluation gives:
``` python
test_epoch = ValidEpoch(
    model=best_model,
    loss=loss,
    metrics=metrics,
    device=DEVICE,
)
 40/40 [00:06<00:00,  7.23it/s, bce_dice_loss - -3.208e+03, iou - 1.163, f-score - 0.9803] 
```

But when I manually score on validation - much much lower result:

``` python 
for i in range(40):
    n = i#np.random.choice(len(valid_dataset))
    image_vis = imread(valid_dataset.images_fps[n])    
    image, gt_mask = valid_dataset[n]
    image, gt_mask = image, gt_mask.transpose(1,2,0)
    gt_mask = gt_mask.squeeze()
    
    x_tensor = torch.from_numpy(image).to(DEVICE).unsqueeze(0)
    pr_mask = best_model.predict(x_tensor)
    pr_mask = (pr_mask.squeeze().cpu().numpy() > 0.5).astype(np.uint8)
    score = iou(torch.from_numpy(gt_mask).float().to(DEVICE), torch.from_numpy(pr_mask).float().to(DEVICE)).data.cpu().numpy().max()
    #rint(score)
    scores.append(score)
    IOU 0.07_
``` 

Then I tried to change preprocessing   to recalling by dividing by 255 - model just learn - IOU near 0.0007 - with almost same result if I measure it manually.

Using Keras achieved almost IOU 0.5 with same almost same pipeline

**Comment by qubvel** [comment]
Hi @Diyago 
Hi did you define loss function and metrics? May be try to set `eps=1`

**Comment by Diyago** [comment]
I have used this way:
```
loss = smp.utils.losses.BCEDiceLoss(eps=1.)
metrics = [
    smp.utils.metrics.IoUMetric(eps=1.),
    smp.utils.metrics.FscoreMetric(eps=1.),
]
```
![Screenshot from 2019-05-26 22-55-34](https://user-images.githubusercontent.com/8073766/58386518-8838de00-8009-11e9-874b-6ba43d3bbdcc.png)

That function gives too high score. I tried to use that function:
``` python
def iou_metric(y_true_in, y_pred_in):
    labels = y_true_in
    y_pred = y_pred_in

    temp1 = np.histogram2d(labels.flatten(), y_pred.flatten(), bins=([0,0.5,1], [0,0.5, 1]))

    intersection = temp1[0]

    area_true = np.histogram(labels,bins=[0,0.5,1])[0]
    area_pred = np.histogram(y_pred, bins=[0,0.5,1])[0]
    area_true = np.expand_dims(area_true, -1)
    area_pred = np.expand_dims(area_pred, 0)

    # Compute union
    union = area_true + area_pred - intersection
  
    # Exclude background from the analysis
    intersection = intersection[1:,1:]
    intersection[intersection == 0] = 1e-9
    
    union = union[1:,1:]
    union[union == 0] = 1e-9

    iou = intersection / union
    return iou
```

It gives about 0 score for images above not 0.4

**Comment by Diyago** [comment]
I rewrote the train loop a bit - now it doesn`t give more 0.5 IOU - but even this number too high

``` python 
# train model for 40 epochs

max_score = 0.4706
train_ious = []
val_ious = []
for i in range(0, 1):
    
    print('\nEpoch: {}'.format(i))
    train_logs = train_epoch.run(train_loader)

    train_iou = score_model(model, train_dataset)
    val_iou = score_model(model,   valid_dataset)
    train_ious.append(train_iou)
    val_ious.append(val_iou)
    
    print('train IOU:', train_iou)
    print('val   IOU:', val_iou)

    # do something (save model, change lr, etc.)
    if max_score < val_iou:
        max_score = val_iou
        torch.save(model, './weights/pytorch/best_model.pth')
        print('Model saved!') 
```

**Comment by Diyago** [comment]
My opinion something wrong with this IOU metric from catalyst project. I have tried second catalyst pipeline as well but with IOU class - same problem

**Comment by qubvel** [comment]
could you share your code?

**Comment by Diyago** [comment]
[pytorch cranes-Copy1.zip](https://github.com/qubvel/segmentation_models.pytorch/files/3224803/pytorch.cranes-Copy1.zip)
Sure

**Comment by Diyago** [comment]
The problem was with the activation function, we need to pass None, because catalyst uses logits
``` python 
loss = smp.utils.losses.BCEDiceLoss(eps=1.)
metrics = [
    smp.utils.metrics.IoUMetric(eps=1., activation = None),
    smp.utils.metrics.FscoreMetric(eps=1.),
]
```
